tutorials/schedule-dataflow-jobs-with-cloud-scheduler/scheduler-dataflow-demo/terraform/main.tf

```diff
@@ -0,0 +1,66 @@
+terraform {
+  backend "gcs" {
+    bucket = "zhong-gcp"
+    prefix = "terraform/state"
+  }
+}
+
+provider "google" {
+  version = "~> 2.20"
+  project = "zhong-gcp"
+}
+
+# Use this data source to get project details. For more information see API.
+# https://www.terraform.io/docs/providers/google/d/google_project.html
+
+data "google_project" "project" {}
+resource "google_cloud_scheduler_job" "scheduler" {
+  name = "scheduler-demo"
+  schedule = "0 0 * * *"
+  # This needs to be us-central1 even if the app engine is in us-central.
+  # You will get a resource not found error if just using us-central.
+  region = "us-central1"
+
+  http_target {
+    http_method = "POST"
+    uri = "https://dataflow.googleapis.com/v1b3/projects/${var.project_id}/locations/${var.region}/templates:launch?gcsPath=gs://zhong-gcp/templates/dataflow-demo-template"
```

Comment:
I'm getting 403 PERMISSION_DENIED from Scheduler when invoking this?  There's a Dataflow Step INFO log for "dataflow.jobs.create" with authorizationInfo: granted: true, which would imply that DF is OK, but no further logs.

Reply:
I updated the doc to use Terraform to create the SA and set the permission. I tried to run the step in a cloud console and it worked for me.

Reply:
Tested, and the Cloud Build SA cannot create/modify/act as other SAs or schedule jobs by default.  Please direct the user to https://cloud.google.com/cloud-build/docs/securing-builds/configure-access-for-cloud-build-service-account#granting_a_role_using_the_iam_page and ask them to add these three roles:

1. Service Accounts > Service Account Admin
2. Service Accounts > Service Account User
3. Cloud Scheduler > Cloud Scheduler Admin

But even after granting all these, assigning the extra roles to the successfully created new SA fails:

Step #0 - "Terraform init": Error: Batch "iam-project-hybrid-prom modifyIamPolicy" for request "Create IAM Members roles/dataflow.admin serviceAccount:scheduler-dataflow-demo@hybrid-prom.iam.gserviceaccount.com for \"project \\\"hybrid-prom\\\"\"" returned error: Error retrieving IAM policy for project "hybrid-prom": googleapi: Error 403: The caller does not have permission, forbidden

...which makes no sense to me since roles/iam.serviceAccountAdmin includes iam.serviceAccounts.getIamPolicy.  terraform graph also shows that the dependency is correctly recognized and the mods only made after the SA exists:

 "[root] google_project_iam_member.cloud-scheduler-dataflow" -> "[root] google_service_account.cloud-scheduler-demo"
 "[root] google_project_iam_member.cloud-scheduler-gcs" -> "[root] google_service_account.cloud-scheduler-demo"

Step #0 - "Terraform init": google_service_account.cloud-scheduler-demo: Creating...
Step #0 - "Terraform init": google_service_account.cloud-scheduler-demo: Creation complete after 2s [id=projects/hybrid-prom/serviceAccounts/scheduler-dataflow-demo@hybrid-prom.iam.gserviceaccount.com]
Step #0 - "Terraform init": google_project_iam_member.cloud-scheduler-dataflow: Creating...
Step #0 - "Terraform init": google_project_iam_member.cloud-scheduler-gcs: Creating...
[errors]

What am I missing? Any ideas?

Reply:
```
Cloud Scheduler Admin
Dataflow Admin
Service Account User
Project IAM Admin
```
Maybe you are missing `Project IAM Admin` Role to manage IAM binding? I reduced the roles to the above four and it worked for me.

Can you try it as well?

Reply:
A-ha!  Project IAM Admin was the missing role, it works now!

Two more tweaks:
+ You need Service Account Admin to create the new SA
- You don't need Dataflow Admin (because it's the new SA that kicks off the job)

Also, in the gcloud builds command you need to use $BUCKET_NAME (not $BUCKET), otherwise it tries to access gs://gs://bucket.

Reply:
I actually removed the `bucket_name` completely. Currently `bucket` represents the bucket name. In the scripts, I have added back the `gs://` prefix for all the bucket paths.